[MLIR][OpenMP] Add Lowering support for OpenMP Declare Mapper directive

TIFitis opened the pull request:
This patch adds HLFIR/FIR lowering support for OpenMP Declare Mapper directive.
Depends on #117045.

TIFitis committed:
[MLIR][OpenMP] Add Lowering support for OpenMP Declare Mapper directive

This patch adds HLFIR/FIR lowering support for OpenMP Declare Mapper directive.

TIFitis committed:
Add emptyline to test.

tblah commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2701,7 +2702,39 @@ static void
+  const auto &mapperName{std::get<std::optional<parser::Name>>(spec.t)};
+  const auto &varType{std::get<parser::TypeSpec>(spec.t)};
+  const auto &varName{std::get<parser::Name>(spec.t)};
+  std::stringstream mapperNameStr;
+  if (mapperName.has_value()) {
+    mapperNameStr << mapperName->ToString();
+  } else {
+    mapperNameStr << "default_"
+                  << varType.declTypeSpec->derivedTypeSpec().name().ToString();
+  }
```
Two nits. Feel free to ignore number 2.
1. Flang **lowering** follows the MLIR style guide, which in this case matches LLVM: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements
2. To me, a `std::stringstream` feels like overkill here. You could use a `std::string` with the concatenation in the else branch handled  by an implicit `Twine` (https://llvm.org/docs/ProgrammersManual.html#llvm-adt-twine-h)

TIFitis replied:
Done.

tblah commented on flang/test/Lower/OpenMP/Todo/omp-declare-mapper.f90:
```diff
@@ -21,7 +21,7 @@ subroutine declare_mapper_1
   real                   :: x, y(nvals)
   !$omp declare mapper (my_type :: var) map (var, var%values (1:var%num_vals))
-!CHECK: not yet implemented: OpenMPDeclareMapperConstruct
+!CHECK: not yet implemented: lowering symbol to HLFIR
```
I'm surprised to see this TODO come up. Please could you fix this before merging so that we can maintain a helpful error message for the user.

TIFitis replied:
This error is now from an unhandled form of map clause rather than declare mapper. As such, I believe it's out of scope for this PR.

I will however subsequently look into fixing it in a separate PR, hope that doesn't hold up this PR.

TIFitis committed:
Addressed reviewer comments. Added assert to check for derived type.

tblah commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2701,7 +2701,42 @@ static void
+  mlir::OpBuilder::InsertPoint insPt = firOpBuilder.saveInsertionPoint();
+  firOpBuilder.setInsertionPointToStart(converter.getModuleOp().getBody());
+  auto mlirType = converter.genType(varType.declTypeSpec->derivedTypeSpec());
+  auto varVal = firOpBuilder.createTemporaryAlloc(
+      converter.getCurrentLocation(), mlirType, varName.ToString());
```
Sorry I didn't notice this before.

So far as I understand, this will create the `fir.alloca` and `hlfir.declare` at the beginning of the MLIR module, not nested in any intermediate operation.

How do you intend to lower this to LLVMIR? We would normally nest these in some kind of "function-like" wrapper operation e.g. `func.func` `fir.global` `omp.private` etc. I wonder if the declare mapper operation needs a nested region for this allocation (like we do for `omp.private` and `omp.declare_reduction`).

TIFitis replied:
From what I understood when trying to implement this lowering: 
The `declMapperOp's` `parentOp` must have a symbol table as it declares a new symbol, this is why I hoisted it to the `ModuleOp` from the `funcOp`.

The `declMapperOp` also relies on a var definition which is immediately used in it's map clause. I can't put this var inside a region, as then it can't be used in the map clause. As such, this new `alloc` is also moved to just before the `declMapperOp`.

Do you have any suggestion to do this in a different way?

tblah replied:
Unfortunately I am not very familiar with the target offload parts of OpenMP so I might have missunderstood something here.

OpenMP 5.2 (section 5.8.8) says
> The visibility and accessibility of this declaration are the same as those of a variable declared at the same location in the program.
Therefore I don't think it is right to be pushing this to the module scope in case there are conflicting mappers declared in different scopes in the same module.

If I understand correctly, the `declare mapper` directive declares `var`:
> The variable declared by var is available for use in all map clauses on the directive, and no part of the variable to be mapped is mapped by default.

So it should be treated as though this parse node is a variable declaration so `createTemporaryAlloc` would be appropriate if this is inside of a function, but if it is not, this should probably be a `fir.global` (like if you declared a module- or file-level variable).

TIFitis replied:
> Therefore I don't think it is right to be pushing this to the module scope in case there are conflicting mappers declared in different scopes in the same module.

It isn't my intention to move the `declMapperOp` to the module scope, but rather MLIR is forcing me to. If there's an alternative I'm missing where I can keep the declMapperOp within the func/block where it was declared, I'd much rather do that. In case there is no way to do that in MLIR, I can prepend the name of the function to the module scope declaration to prevent conflicts.

> So it should be treated as though this parse node is a variable declaration so `createTemporaryAlloc` would be appropriate if this is inside of a function, but if it is not, this should probably be a `fir.global` (like if you declared a module- or file-level variable).

If we go forward with keeping the `declMapperOp` in the module scope, then I'll alter the `var` declaration to be a `fir.global`.

kiranchandramohan replied:
Isn't `declare mapper` providing a specification which is to be used for mapping in the associated regions? The variable in the `declare mapper` has its scope inside the construct since it is only used for specifying how to map. Similar to what @tblah is saying, you could do something like the following.

Assuming %x is the variable that is being mapped by the rules of the declare mapper:
```
omp.declare_mapper @default_my_type {
^bb0(%[[VAR:.*]]: [[TYPE]]):
  %v1 = omp.bounds
  %v2 = omp.map.info var_ptr(%[[VAR:.*]])
  omp.yield %v2
}

%x_map = omp.map.info var_ptr(%x) decl_map(@default_my_type)
omp.target_enter_data   map_entries(%x_map)
```

or 

```
omp.declare_mapper @default_my_type {
^bb0(%[[VAR:.*]]: [[TYPE]]):
  %v1 = omp.bounds
  %v2 = omp.map.info var_ptr(%[[VAR:.*]])
  omp.yield %v2
}
omp.target_enter_data   map_entries(%x : @default_my_type)
```

The alternative would be to not have a `declare mapper` construct and apply the declare mapper during lowering itself. But this would be early lowering and we wont be able to share code with Clang as well.

TIFitis replied:
From what I understand, declare mapper directive doesn't simply apply those mapping rules as is to `%x` in your example, or at least not in the plain manner suggested.

A new mapper function is emitted, and a pointer to this function is passed on to the runtime kernel call for target. The OMPRTL takes offload_mappers as the last argument, which is an array of pointers to the user defined mapper funcs.

If we go with the approach you are suggesting above, we would have to decouple the `map clause` from the declare mapper directive and have floating `mapInfoOps` in its nested region. This seems like an off approach to me. Also, we would need to handle this as a special case everywhere we handle regular map clauses, including the several lowering stages and passes like `mapInfoFinalization`.

If we really want to enforce the scoping rule for the declare mapper `var`, perhaps we can do so in the verifier to check that `%var` has a single use, i.e, the `declMapperOp`. Let me know if that sounds reasonable and we can leave the `%var` declaration just above the `declMapperOp`.

As for the placement of the `declMapperOp` at the Module level or Func level, I am undecided. The `declMapperOp` is in some sense akin to a function declaration and will be lowered into one eventually in the llvm IR. We're just delaying that process until we lower to llvm IR, as it saves us unnecessary work, and also we can share the lowering between Clang and MLIR through the OMPIRBUilder.

With that being said, as we are not planning on expanding it into a function at the MLIR level, I am not opposed to leaving it inside the func it's defined in to retain scoping rules. We can do so by changing the `declMapperOp` to no longer declare a symbol for itself, to prevent MLIR from complaining about `ModuleOp` not being it's parent.
This would mean that later when lowering map clauses, instead of doing a sym_table lookup for the mapper, we would have to scan the code for `declMapperOps` instead and match. This may be slightly inefficient, but I don't think should be too much of an issue.

Let me know what are your thoughts on this and how you'd like me to proceed :)

kiranchandramohan replied:
Looks like @skatrak is also thinking similarly, see his comment in https://github.com/llvm/llvm-project/pull/117045#discussion_r1858907868

> A new mapper function is emitted, and a pointer to this function is passed on to the runtime kernel call for target. The OMPRTL takes offload_mappers as the last argument, which is an array of pointers to the user defined mapper funcs.

If this is the way it is implemented in Clang, I think you can do it this way by creating the mapper function from the MLIR region-based declare mapper specification operation.

TIFitis replied:
@agozillon Do you have any insight or suggestion to handling the declMapperOp map clause one way or the other?

agozillon replied:
I don't have much insight into how the (soon to be) declare mapper portion of the OMPIRBuilder or omp.private/declare_reduction work unfortunately, and I am nowhere near an MLIR guru, still at a beginner level I'd wager, so take my comment with a grain of salt :-) 

I do think it would be ideal to maintain the correct scoping if we can, as the specification seems to be quite clear on it, although, I imagine if there is significant difficulties in doing so at an MLIR level as long as semantically it is the same to a user (i.e. no issues / bugs), then the specification doesn't really have room to complain as we're in the land of implementation details. However, if we can maintain the scoping as it seems to be suggested is possible, that'd be awesome :-)

Otherwise, I do quite like the lambda-esque approach conceptually as it does seem to achieve the desired results as far as scoping goes, and it is tidy in the sense that it can be shared across multiple mappings of the variable/type the mapper has been declared for quite trivially and follows the whole concept that declare mapper is effectively creating a specification/template of the desired mapping. Saying that, I am not against the simpler method either, I am not sure which would be easier to perform transformations/optimisations on if we ever wished to. 

As far as the lowering goes, I don't think it'd be too difficult in either case. The main pass that does things with map information is as you've mentioned MapInfoFinalization, which I believe should only need minor tweaks, primarily to recognize declare mappers as something that contains map.info that needs walked over, you would also likely have to teach it to add any new maps it generates to the yield operation as well. So some tedium but nothing too egregious I think, and more than happy to help with it. I do not think there's anything incredibly complex as far as the OpenMP -> OpenMP dialect translation that occurs when we lower from FIR to LLVM dialect goes as far as the map info, so it'd likely just be a case of mimic'ng how other OMP operations that follow the same format work at that stage. For MLIR -> LLVM-IR, you'd likely still have to do a lot of the same legwork in either case with some differences I would imagine, as at the end of the day the map info and bounds information is just an information vehicle, it'll have to be converted into whatever the declare mapper portion of the OMPIRBuilder desires, which won't be the exact same as what use_device_addr does or what map does. 

But as with anything complex, there will most definitely be issues and unforeseen things you'll likely run into and I am not the soldier in the trenches so to speak, so I imagine you'll be able to assess the difficulties better than I can :-)  

P.S. I would love to see what Clang emits for a declare mapper in the IR if you have a spare blob of IR handy or a small sample program I can compile! Might help to give some further insight into what the end result for us will be (at least it'll help me in any case, will give me a better idea of what the expectations for what the MLIR -> LLVM-IR portion will need to do).

kiranchandramohan replied:
> Also, I'm not sure what we stand to gain from this approach. 

In the approach that you are proposing, there is a variable and map.info that is created in the function scope with an alloca. This is not correct since the existence of a declare mapper does not cause the creation of a variable or a mapping.

Worst case, 
-> this can appear to be aliasing with other memory locations and reduce performance. 
-> increase live-ins of openmp regions when outlined

> Every other instance of where a directive has a mapClause associated with it, does so by creating the map related Ops just above it. Why are we nesting them in case of declMapperOp?

Because it is a declaration. It will only have an effect when it is used either via its usage in implicit/default mapping of the derived type for which it was created or when it is directly referred to by a map(mapper...) clause.

I think you have a reasonable point here because of the following:
```
1. If a DECLARE MAPPER directive is not specified for a type DT, a predefined mapper exists for type DT as if the type DT had appeared in the directive as follows:

!$OMP DECLARE MAPPER (DT :: var) MAP (TOFROM: var)
2. If a variable is not a scalar then it is treated as if it had appeared in a map clause with a map-type of tofrom. Which is effectively equivalent to the following and extending declare mapper for non-derived types:
!$OMP DECLARE MAPPER (T :: var) MAP (TOFROM: var)
```
We can thus argue that all mapping should also generate declarations that are ultimately all handled in a separate pass in flang/mlir. May be 1 and 2 are trivial and that is why it is not done that way.

kiranchandramohan replied:
Besides the representation, we should also clarify 
-> where we create the map_entries for the relevant operations (like target) for which the declare mapper implicitly applies
-> where the composition of map-types (map-type decay) from the map clause of declare mapper and the map clause of the relevant operation (like target) happens

Note: All these are for discussion and should be treated as suggestions.

agozillon replied:
> ```
> 1. If a DECLARE MAPPER directive is not specified for a type DT, a predefined mapper exists for type DT as if the type DT had appeared in the directive as follows:
> 
> !$OMP DECLARE MAPPER (DT :: var) MAP (TOFROM: var)
> 2. If a variable is not a scalar then it is treated as if it had appeared in a map clause with a map-type of tofrom. Which is effectively equivalent to the following and extending declare mapper for non-derived types:
> !$OMP DECLARE MAPPER (T :: var) MAP (TOFROM: var)
> ```

I think the keyword here is likely "as if", so as long as the effects are as described it's reasonable would be my reading, and if we really wanted to be exact about the wording we'd generate/embed our own equivalent pragmas to the above for all default mappings, and then lower them, so not just at the MLIR level. However, saying that I am not against defining a default declare mapper for all cases once it's in place,  it might tidy things up a bit, but it may also be more complicated/trouble than it's worth, in either case I am fine with the approach of defining default declare mappers if we'd like to go down that route :-) 

I'd also love it if whatever implementation we landed on was compatible with the OpenACC implementations documentation/approach to mapping descriptors via runtime calls, as I'd like to move towards that eventually when I have some time to dig into it and see if it's viable for us. I imagine it will be, I just don't know a ton about the region'd approach so hope it wouldn't be prohibitive of this.

TIFitis replied:
I have made the required changes to both PRs to update the representation for the DeclMapperOp.

Some changes I've incorporated are:

Inside DeclMapperOp's region I've introduce a new `DeclMapperInfoOp` to which I've attached the `MapClause`. Not having any `MapClause` explicitly associated seemed weird to me, also walking through the region collecting all the `MapInfoOps` for processing in various places in the code base seemed like a bad design to me.

Also, instead of using the block arg, I've created a `temporary alloca` inside the region. This is to save the hassle of binding the block_arg to the symbol, binding the `alloca` is much more straightforward.

We can drop the entire `DeclMapperOp` including the region once it reaches `OpenMPTOLLVMIRTranslation`.

Let me what are your thoughts on this approach :)

TIFitis replied:
@kiranchandramohan @tblah @skatrak @agozillon Any thoughts on this approach?

kiranchandramohan replied:
You should feel free to proceed with the approval of @agozillon and @skatrak. You can set up a quick call with them and describe your approach and come to a conclusion. You can report the outcomes somewhere in this patch or write in discourse. 

> Inside DeclMapperOp's region I've introduce a new DeclMapperInfoOp to which I've attached the MapClause. Not having any MapClause explicitly associated seemed weird to me, also walking through the region collecting all the MapInfoOps for processing in various places in the code base seemed like a bad design to me.

The idea would generally be to inline the whole declare mapper operation region, replacing the block_arg with the real variable that is going to be mapped. Would an alloca always be created for all kinds of mappings? If not committing to an alloca might mean you have to delete it in some circumstances.

> We can drop the entire DeclMapperOp including the region once it reaches OpenMPTOLLVMIRTranslation.

As in just drop it without using it? Or using it create the `@.omp_mapper._ZTS1T.deep_copy` function for the declare mapper (`#pragma omp declare mapper(deep_copy : T abc) map(abc, abc.ptr[ : abc.buf_size])`) in the C example you gave below.

You have not talked about the following two points.
```
-> where we create the map_entries for the relevant operations (like target) for which the declare mapper implicitly applies. Currently, this is done during lowering. But in this patch you have solely focused on creating the declare mapper.
-> where the composition of map-types (map-type decay) from the map clause of declare mapper and the map clause of the relevant operation (like target) happens
```

Couple of other things that came to my mind :
-> Since declare mappers are in the specification section, it can also occur in modules. We have not added code to propagate it to the module file. If this is not urgent for you, I can fix it sometime.
-> You should test the case where the declare mapper is in the host subroutine
```
subroutine A
type t
end type
declare mapper(t)
contains
subroutine B
!$omp target 
! use a var of type t
!$omp end target
end subroutine
```

TIFitis replied:
> You should feel free to proceed with the approval of @agozillon and @skatrak.
Thanks for the go ahead, I'll consult with them before proceeding, and post any relevant findings here.

> > Inside DeclMapperOp's region I've introduce a new DeclMapperInfoOp to which I've attached the MapClause. Not having any MapClause explicitly associated seemed weird to me, also walking through the region collecting all the MapInfoOps for processing in various places in the code base seemed like a bad design to me.
> 
> The idea would generally be to inline the whole declare mapper operation region, replacing the block_arg with the real variable that is going to be mapped. Would an alloca always be created for all kinds of mappings? If not committing to an alloca might mean you have to delete it in some circumstances.

In case of DeclareMapper the mapping is handled in a ad-hoc basis. There is no real variable to be replaced later. The variable simply exists as a dummy to represent the mapping information for other real variables where the mapping would be used. However, this mapping isn't applied directly to those variables either, rather the runtime is notified that a mapper function exists for these variables.

> > We can drop the entire DeclMapperOp including the region once it reaches OpenMPTOLLVMIRTranslation.
> 
> As in just drop it without using it? Or using it create the `@.omp_mapper._ZTS1T.deep_copy` function for the declare mapper (`#pragma omp declare mapper(deep_copy : T abc) map(abc, abc.ptr[ : abc.buf_size])`) in the C example you gave below.
Yes, I mean't each declareMapperOp would be resolved to a function like `@.omp_mapper._ZTS1T.deep_copy`. By dropping it, I mean't it's region won't be separately lowered.

> You have not talked about the following two points.
> 
> ```
> -> where we create the map_entries for the relevant operations (like target) for which the declare mapper implicitly applies. Currently, this is done during lowering. But in this patch you have solely focused on creating the declare mapper.
> -> where the composition of map-types (map-type decay) from the map clause of declare mapper and the map clause of the relevant operation (like target) happens
> ```

I'll address the implicit mapping either in the DeclMapper lowering support for the mapClause or a separate PR soon after that.

The composition of mapClause AFAIK is likely handled in the OMPIRBuilder. If that's not the case, I'll address it in a future patch. 

> Couple of other things that came to my mind : -> Since declare mappers are in the specification section, it can also occur in modules. We have not added code to propagate it to the module file. If this is not urgent for you, I can fix it sometime. 

Please feel free to do this if you're already familiar with what needs to be done, as I'm not. Otherwise, I can add it to my TODO list :)

-> You should test the case where the declare mapper is in the host subroutine
> 
> ```
> subroutine A
> type t
> end type
> declare mapper(t)
> contains
> subroutine B
> !$omp target 
> ! use a var of type t
> !$omp end target
> end subroutine
> ```

Thanks, I'll test this in the mapClause mapper support PR once I have it ready.

Many thanks for the review.

TIFitis committed:
Add lowering changes for declMapperOp's region.

TIFitis commented:
As @agozillon requested, here's a sample of how Clang lowers declare mapper.

**C Code:**
```
typedef struct {
  int *ptr;
  int buf_size;
} T;
#pragma omp declare mapper(deep_copy : T abc) map(abc, abc.ptr[ : abc.buf_size])
int main() {
  T xyz;
#pragma omp target data map(mapper(deep_copy), tofrom : xyz)
  {
  }
  return 0;
}
```

**LLVM IR:**
```
@.offload_sizes = private unnamed_addr constant [1 x i64] [i64 16]
@.offload_maptypes = private unnamed_addr constant [1 x i64] [i64 3]

; Function Attrs: noinline nounwind optnone uwtable
define dso_local i32 @main() #0 {
entry:
  %retval = alloca i32, align 4
  %xyz = alloca %struct.T, align 8
  %.offload_baseptrs = alloca [1 x ptr], align 8
  %.offload_ptrs = alloca [1 x ptr], align 8
  %.offload_mappers = alloca [1 x ptr], align 8
  store i32 0, ptr %retval, align 4
  %0 = getelementptr inbounds [1 x ptr], ptr %.offload_baseptrs, i32 0, i32 0
  store ptr %xyz, ptr %0, align 8
  %1 = getelementptr inbounds [1 x ptr], ptr %.offload_ptrs, i32 0, i32 0
  store ptr %xyz, ptr %1, align 8
  %2 = getelementptr inbounds [1 x ptr], ptr %.offload_mappers, i64 0, i64 0
  store ptr @.omp_mapper._ZTS1T.deep_copy, ptr %2, align 8
  %3 = getelementptr inbounds [1 x ptr], ptr %.offload_baseptrs, i32 0, i32 0
  %4 = getelementptr inbounds [1 x ptr], ptr %.offload_ptrs, i32 0, i32 0
  call void @__tgt_target_data_begin_mapper(ptr @1, i64 -1, i32 1, ptr %3, ptr %4, ptr @.offload_sizes, ptr @.offload_maptypes, ptr null, ptr %.offload_mappers)
  %5 = getelementptr inbounds [1 x ptr], ptr %.offload_baseptrs, i32 0, i32 0
  %6 = getelementptr inbounds [1 x ptr], ptr %.offload_ptrs, i32 0, i32 0
  call void @__tgt_target_data_end_mapper(ptr @1, i64 -1, i32 1, ptr %5, ptr %6, ptr @.offload_sizes, ptr @.offload_maptypes, ptr null, ptr %.offload_mappers)
  ret i32 0
}

; Function Attrs: noinline nounwind uwtable
define internal void @.omp_mapper._ZTS1T.deep_copy(ptr noundef %0, ptr noundef %1, ptr noundef %2, i64 noundef %3, i64 noundef %4, ptr noundef %5) #1 {
entry:
  %.addr = alloca ptr, align 8
  %.addr1 = alloca ptr, align 8
  %.addr2 = alloca ptr, align 8
  %.addr3 = alloca i64, align 8
  %.addr4 = alloca i64, align 8
  %.addr5 = alloca ptr, align 8
  store ptr %0, ptr %.addr, align 8
  store ptr %1, ptr %.addr1, align 8
  store ptr %2, ptr %.addr2, align 8
  store i64 %3, ptr %.addr3, align 8
  store i64 %4, ptr %.addr4, align 8
  store ptr %5, ptr %.addr5, align 8
  %6 = load i64, ptr %.addr3, align 8
  %7 = load ptr, ptr %.addr, align 8
  %8 = load ptr, ptr %.addr1, align 8
  %9 = load ptr, ptr %.addr2, align 8
  %10 = udiv exact i64 %6, 16
  %11 = getelementptr %struct.T, ptr %9, i64 %10
  %12 = load i64, ptr %.addr4, align 8
  %13 = load ptr, ptr %.addr5, align 8
  %omp.arrayinit.isarray = icmp sgt i64 %10, 1
  %14 = and i64 %12, 8
  %15 = icmp ne ptr %8, %9
  %16 = and i64 %12, 16
  %17 = icmp ne i64 %16, 0
  %18 = and i1 %15, %17
  %19 = or i1 %omp.arrayinit.isarray, %18
  %.omp.array..init..delete = icmp eq i64 %14, 0
  %20 = and i1 %19, %.omp.array..init..delete
  br i1 %20, label %.omp.array..init, label %omp.arraymap.head

.omp.array..init:                                 ; preds = %entry
  %21 = mul nuw i64 %10, 16
  %22 = and i64 %12, -4
  %23 = or i64 %22, 512
  call void @__tgt_push_mapper_component(ptr %7, ptr %8, ptr %9, i64 %21, i64 %23, ptr %13)
  br label %omp.arraymap.head

omp.arraymap.head:                                ; preds = %.omp.array..init, %entry
  %omp.arraymap.isempty = icmp eq ptr %9, %11
  br i1 %omp.arraymap.isempty, label %omp.done, label %omp.arraymap.body

omp.arraymap.body:                                ; preds = %omp.type.end19, %omp.arraymap.head
  %omp.arraymap.ptrcurrent = phi ptr [ %9, %omp.arraymap.head ], [ %omp.arraymap.next, %omp.type.end19 ]
  %ptr = getelementptr inbounds nuw %struct.T, ptr %omp.arraymap.ptrcurrent, i32 0, i32 0
  %ptr6 = getelementptr inbounds nuw %struct.T, ptr %omp.arraymap.ptrcurrent, i32 0, i32 0
  %24 = load ptr, ptr %ptr6, align 8
  %arrayidx = getelementptr inbounds nuw i32, ptr %24, i64 0
  %buf_size = getelementptr inbounds nuw %struct.T, ptr %omp.arraymap.ptrcurrent, i32 0, i32 1
  %25 = load i32, ptr %buf_size, align 8
  %conv = sext i32 %25 to i64
  %26 = mul nuw i64 %conv, 4
  %27 = getelementptr %struct.T, ptr %omp.arraymap.ptrcurrent, i32 1
  %28 = ptrtoint ptr %27 to i64
  %29 = ptrtoint ptr %omp.arraymap.ptrcurrent to i64
  %30 = sub i64 %28, %29
  %31 = sdiv exact i64 %30, ptrtoint (ptr getelementptr (i8, ptr null, i32 1) to i64)
  %32 = call i64 @__tgt_mapper_num_components(ptr %7)
  %33 = shl i64 %32, 48
  %34 = add nuw i64 0, %33
  %35 = and i64 %12, 3
  %36 = icmp eq i64 %35, 0
  br i1 %36, label %omp.type.alloc, label %omp.type.alloc.else

omp.type.alloc:                                   ; preds = %omp.arraymap.body
  %37 = and i64 %34, -4
  br label %omp.type.end

omp.type.alloc.else:                              ; preds = %omp.arraymap.body
  %38 = icmp eq i64 %35, 1
  br i1 %38, label %omp.type.to, label %omp.type.to.else

omp.type.to:                                      ; preds = %omp.type.alloc.else
  %39 = and i64 %34, -3
  br label %omp.type.end

omp.type.to.else:                                 ; preds = %omp.type.alloc.else
  %40 = icmp eq i64 %35, 2
  br i1 %40, label %omp.type.from, label %omp.type.end

omp.type.from:                                    ; preds = %omp.type.to.else
  %41 = and i64 %34, -2
  br label %omp.type.end

omp.type.end:                                     ; preds = %omp.type.from, %omp.type.to.else, %omp.type.to, %omp.type.alloc
  %omp.maptype = phi i64 [ %37, %omp.type.alloc ], [ %39, %omp.type.to ], [ %41, %omp.type.from ], [ %34, %omp.type.to.else ]
  call void @__tgt_push_mapper_component(ptr %7, ptr %omp.arraymap.ptrcurrent, ptr %omp.arraymap.ptrcurrent, i64 %31, i64 %omp.maptype, ptr null)
  %42 = add nuw i64 281474976710659, %33
  %43 = and i64 %12, 3
  %44 = icmp eq i64 %43, 0
  br i1 %44, label %omp.type.alloc7, label %omp.type.alloc.else8

omp.type.alloc7:                                  ; preds = %omp.type.end
  %45 = and i64 %42, -4
  br label %omp.type.end12

omp.type.alloc.else8:                             ; preds = %omp.type.end
  %46 = icmp eq i64 %43, 1
  br i1 %46, label %omp.type.to9, label %omp.type.to.else10

omp.type.to9:                                     ; preds = %omp.type.alloc.else8
  %47 = and i64 %42, -3
  br label %omp.type.end12

omp.type.to.else10:                               ; preds = %omp.type.alloc.else8
  %48 = icmp eq i64 %43, 2
  br i1 %48, label %omp.type.from11, label %omp.type.end12

omp.type.from11:                                  ; preds = %omp.type.to.else10
  %49 = and i64 %42, -2
  br label %omp.type.end12

omp.type.end12:                                   ; preds = %omp.type.from11, %omp.type.to.else10, %omp.type.to9, %omp.type.alloc7
  %omp.maptype13 = phi i64 [ %45, %omp.type.alloc7 ], [ %47, %omp.type.to9 ], [ %49, %omp.type.from11 ], [ %42, %omp.type.to.else10 ]
  call void @__tgt_push_mapper_component(ptr %7, ptr %omp.arraymap.ptrcurrent, ptr %omp.arraymap.ptrcurrent, i64 16, i64 %omp.maptype13, ptr null)
  %50 = add nuw i64 281474976710675, %33
  %51 = and i64 %12, 3
  %52 = icmp eq i64 %51, 0
  br i1 %52, label %omp.type.alloc14, label %omp.type.alloc.else15

omp.type.alloc14:                                 ; preds = %omp.type.end12
  %53 = and i64 %50, -4
  br label %omp.type.end19

omp.type.alloc.else15:                            ; preds = %omp.type.end12
  %54 = icmp eq i64 %51, 1
  br i1 %54, label %omp.type.to16, label %omp.type.to.else17

omp.type.to16:                                    ; preds = %omp.type.alloc.else15
  %55 = and i64 %50, -3
  br label %omp.type.end19

omp.type.to.else17:                               ; preds = %omp.type.alloc.else15
  %56 = icmp eq i64 %51, 2
  br i1 %56, label %omp.type.from18, label %omp.type.end19

omp.type.from18:                                  ; preds = %omp.type.to.else17
  %57 = and i64 %50, -2
  br label %omp.type.end19

omp.type.end19:                                   ; preds = %omp.type.from18, %omp.type.to.else17, %omp.type.to16, %omp.type.alloc14
  %omp.maptype20 = phi i64 [ %53, %omp.type.alloc14 ], [ %55, %omp.type.to16 ], [ %57, %omp.type.from18 ], [ %50, %omp.type.to.else17 ]
  call void @__tgt_push_mapper_component(ptr %7, ptr %ptr, ptr %arrayidx, i64 %26, i64 %omp.maptype20, ptr null)
  %omp.arraymap.next = getelementptr %struct.T, ptr %omp.arraymap.ptrcurrent, i32 1
  %omp.arraymap.isdone = icmp eq ptr %omp.arraymap.next, %11
  br i1 %omp.arraymap.isdone, label %omp.arraymap.exit, label %omp.arraymap.body

omp.arraymap.exit:                                ; preds = %omp.type.end19
  %omp.arrayinit.isarray21 = icmp sgt i64 %10, 1
  %58 = and i64 %12, 8
  %.omp.array..del..delete = icmp ne i64 %58, 0
  %59 = and i1 %omp.arrayinit.isarray21, %.omp.array..del..delete
  br i1 %59, label %.omp.array..del, label %omp.done

.omp.array..del:                                  ; preds = %omp.arraymap.exit
  %60 = mul nuw i64 %10, 16
  %61 = and i64 %12, -4
  %62 = or i64 %61, 512
  call void @__tgt_push_mapper_component(ptr %7, ptr %8, ptr %9, i64 %60, i64 %62, ptr %13)
  br label %omp.done

omp.done:                                         ; preds = %.omp.array..del, %omp.arraymap.exit, %omp.arraymap.head
  ret void
}
```

TIFitis commented:
@kiranchandramohan I discussed the current approach with @skatrak today. When trying to implement the mapper lowering for the map clause, it became apparent that we need to add the `declMapperOp` name to the `SymbolTable`. As such, we would also need to hoist the `declareMapperOp` to the `ModuleOp`.

I am however struggling with name mangling the `declMapperOp` such that scoping information is preserved, and we don't have clashing `declMapperOps` from different nested scopes with the same name.

Take the following example:
```
program my_prog
   type my_type
      integer               :: num
   end type
   !$omp declare mapper (my_mapper : my_type :: my_var) map (my_var)
contains
   subroutine test
      type(my_type)            :: xyz
      !$omp target enter data map(mapper(my_mapper), to: xyz)
   end subroutine test
end program my_prog
```

Here after mangling the, `declMapperOp` symbol name becomes `_QQFmy_mapper`. But when trying to mangle the name that occurs in the mapClause before lookup results in `_QQFFtestmy_mapper`.

Do you know any mechanism in Fortran lowering that could help resolve this issue?

skatrak commented:
After the discussion with @TIFitis yesterday, I think the overall approach currently proposed makes sense. I would just like to share a few related thoughts not for this PR stack, but rather some longer term potential improvements that might make sense to do:
- I think the new `omp.declare_mapper_info` operation currently helps not having to introduce too many new features at once, and rely on the existing infrastructure as much as possible, but I believe eventually this should be replaced with an `omp.yield` of the necessary `omp.map.info` operations and potentially some changes to the verification and handling of `map` clauses.
- Long-term I was thinking that we may want to represent type-dependent handling of `map` clauses in a way more similar to what we do for privatization. Initially I was thinking that we could use the `omp.declare_mapper` operation for this, but it's true that it's only intended to be used for derived types / structures. However, my thinking is that if we introduced something like an `omp.map` operation holding region(s) that define mapping for a given type of variable, we may be able to avoid the need for fix-up passes to add new mapped variables for e.g. allocatables or other complex types. Ideally, it would be able to integrate with `omp.declare_mapper` operations and unify map-related handling so that there's fewer corner cases or at least centralize that kind of logic. This is only a high-level idea at this point, not something I have actually spent time thinking through in detail, but perhaps something worth exploring at some point.

kparzysz commented:
> Are the symbols not being resolved properly? Or did I miss a point?

I'm working on something different at the moment, but for debugging purposes I changed the tree dumper to show symbol information too.  Here's the output for this code.  It shows that the variable `my_mapper` in the declare target, and in the map clause later on use the same symbol: 0x61f689c35c30.
```
Program -> ProgramUnit -> MainProgram
| ProgramStmt -> Name = '(0x61f689c354b0) [my_prog: MainProgram]'
| SpecificationPart
| | ImplicitPart ->
| | DeclarationConstruct -> SpecificationConstruct -> DerivedTypeDef
| | | DerivedTypeStmt
| | | | Name = '(0x61f689c35870) [my_type: DerivedType components: num]'
| | | ComponentDefStmt -> DataComponentDefStmt
| | | | DeclarationTypeSpec -> IntrinsicTypeSpec -> IntegerTypeSpec ->
| | | | ComponentOrFill -> ComponentDecl
| | | | | Name = '(0x61f689c35a50) [num size=4 offset=0: ObjectEntity type: INTEGER(4)]'
| | | EndTypeStmt ->
| | DeclarationConstruct -> SpecificationConstruct -> OpenMPDeclarativeConstruct -> OpenMPDeclareMapperConstruct
| | | Verbatim
| | | OmpDeclareMapperSpecifier
| | | | Name = '(0x61f689c35c30) [my_mapper: Misc ConstructName]'
| | | | TypeSpec -> DerivedTypeSpec
| | | | | Name = '(0x61f689c35870) [my_type: DerivedType components: num]'
| | | | Name = '(0x61f689c35e10) [my_var (OmpMapToFrom) size=4 offset=0: ObjectEntity type: TYPE(my_type)]'
| | | OmpClauseList -> OmpClause -> Map -> OmpMapClause
| | | | OmpObjectList -> OmpObject -> Designator -> DataRef -> Name = '(0x61f689c35e10) [my_var (OmpMapToFrom) size=4 offset=0: ObjectEntity type: TYPE(my_type)]'
| | | | bool = 'true'
| ExecutionPart -> Block
| InternalSubprogramPart
| | ContainsStmt
| | InternalSubprogram -> SubroutineSubprogram
| | | SubroutineStmt
| | | | Name = '(0x61f689c35690) [test (Subroutine): Subprogram ()]'
| | | SpecificationPart
| | | | ImplicitPart ->
| | | | DeclarationConstruct -> SpecificationConstruct -> TypeDeclarationStmt
| | | | | DeclarationTypeSpec -> Type
| | | | | | DerivedTypeSpec
| | | | | | | Name = '(0x61f689c35870) [my_type: DerivedType components: num]'
| | | | | EntityDecl
| | | | | | Name = '(0x61f689c361d0) [xyz (OmpMapTo) size=4 offset=0: ObjectEntity type: TYPE(my_type)]'
| | | ExecutionPart -> Block
| | | | ExecutionPartConstruct -> ExecutableConstruct -> OpenMPConstruct -> OpenMPStandaloneConstruct -> OpenMPSimpleStandaloneConstruct
| | | | | OmpSimpleStandaloneDirective -> llvm::omp::Directive = target enter data
| | | | | OmpClauseList -> OmpClause -> Map -> OmpMapClause
| | | | | | Modifier -> OmpMapper -> Name = '(0x61f689c35c30) [my_mapper: Misc ConstructName]'
| | | | | | Modifier -> OmpMapType -> Value = To
| | | | | | OmpObjectList -> OmpObject -> Designator -> DataRef -> Name = '(0x61f689c361d0) [xyz (OmpMapTo) size=4 offset=0: ObjectEntity type: TYPE(my_type)]'
| | | | | | bool = 'true'
| | | EndSubroutineStmt -> Name = 'test'
| EndProgramStmt -> Name = 'my_prog'
```

TIFitis commented:
@kiranchandramohan @kparzysz I guess I must be using the name mangler in an incorrect way then. I've added the code snippets I am using when lowering and later performing a lookup. Please let me know what would be the correct way of doing this.

When lowering DeclMapperOp:
```
static void
genOMP(lower::AbstractConverter &converter, lower::SymMap &symTable,
       semantics::SemanticsContext &semaCtx, lower::pft::Evaluation &eval,
       const parser::OpenMPDeclareMapperConstruct &declareMapperConstruct) {
  fir::FirOpBuilder &firOpBuilder = converter.getFirOpBuilder();
  const auto &spec =
      std::get<parser::OmpDeclareMapperSpecifier>(declareMapperConstruct.t);
  const auto &mapperName{std::get<std::optional<parser::Name>>(spec.t)};
  std::string mapperNameStr;
  if (mapperName.has_value())
    mapperNameStr = mapperName->ToString();
  else
    mapperNameStr =
        "default_" + varType.declTypeSpec->derivedTypeSpec().name().ToString();

  // This is my current implementation.
  // It returns _QQFmy_mapper as mapperNameStr.
  mapperNameStr = converter.mangleName(*mapperName->symbol);

  auto declMapperOp = firOpBuilder.create<mlir::omp::DeclareMapperOp>(
      loc, mapperNameStr, mlirType);
  converter.getMLIRSymbolTable()->insert(declMapperOp);

  // This code produces the error:
  // LLVM ERROR: /home/akash/Documents/llvm-project/flang/lib/Lower/Mangler.cpp:184: not yet implemented: symbol mangling
  // mapperNameStr = converter.mangleName(*mapperName->symbol);
  
  // This code produces the error:
  // ** symbol not properly mapped **
  // symTable.lookupSymbol(mapperName->symbol).dump();
```

When trying to lookup DeclareMapper from the mapClause mapper:

```
      auto mapperIdName = mappers->front().v.id().symbol->name().ToString();
      mapperIdName = converter.mangleName(mapperIdName);

      // Here mapperIdName returns _QQFFtestmy_mapper which fails the
      // lookup inside the asser.

      // assert(converter.getMLIRSymbolTable()->lookup(mapperIdName) &&
      //        "mapper not found");
      mapperId = mlir::FlatSymbolRefAttr::get(&converter.getMLIRContext(),
                                              mapperIdName);
```

kparzysz commented:
1. If you want to create an MLIR op for the mapper, you can give it some name (you can mangle it if you want).  You will need to store it in the MLIR somewhere/somehow.
2. If you want to record the use of a mapper in a clause, you can just use the name of the mapper in the same form as in (1), i.e. you decide how to generate the name here.
3. If you want to actually apply the mapper, you need to find it, but it's up to you how you do it: if you're looking for it in the MLIR module then your code must have put it there, if you're looking for it in the AST then you can find the declaration of the mapper by the symbol associated with its name.

kparzysz commented:
I looked at the clang code and it's not clear to me why they do it at runtime.  Mappers can refer to other mappers (for sub-objects), but all mappers and all type layouts should be present in (or obtainable from) the AST[1].  In other words, clang should be able to emit all the "expanded" map clauses (i.e. after application of the mappers) at compile-time.  It's possible that instead of emitting all that code inline, they put it in a runtime function to save code size.

[1] The mapper definition must be visible at the time of use in a clause, so all mappers (recursively) should be visible.

TIFitis committed:
Add required changes to hoist DeclareMapperOp to the ModuleOp's region.
Add a new name mangling method which accepts a user supplied scope to mangle the name, in addition to the existing function which always used the currentScope.

TIFitis committed:
Update MapInfoFinalization pass to look for MapInfoOps inside DeclareMapperOps as well as inside the Module.

TIFitis commented:
@kiranchandramohan @skatrak @agozillon Hi everyone! I was on vacation these last couple of weeks, apologies for the delay.

I have added PR #124746 with which we have all the implementation in place for declare mappers.

It would be great to resume the review process for this PR stack.

Thanks :)

Note: I'll add a couple of patches to support the implicit default mapping support for declare mappers.

TIFitis commented:
Polite request for review 🙂

tblah commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2612,7 +2612,54 @@ static void
+        mapperNameStr, *varType.declTypeSpec->derivedTypeSpec().GetScope());
+  }
+
+  // Save insert point just after the DeclMapperOp.
+  mlir::OpBuilder::InsertPoint insPt = firOpBuilder.saveInsertionPoint();
```
```suggestion
  // Save current insertion point before moving to the module scope to create the DeclareMapperOp
  mlir::OpBuilder::InsertionGuard guard(builder);
```

I don't think it makes sense to say the insert point is before or after the DeclMapperOp because the DeclMapperOp will be at module scope. I've also changed to use the insertion point guard because it is less error prone if a conditional early return is added later.

TIFitis replied:
Done :)

tblah commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2612,7 +2612,54 @@ static void
+  ClauseProcessor cp(converter, semaCtx, clauses);
+  cp.processMap(loc, stmtCtx, clauseOps);
+  firOpBuilder.create<mlir::omp::DeclareMapperInfoOp>(loc, clauseOps.mapVars);
+
+  // Restore the insert point to just after the DeclareMapperOp.
+  firOpBuilder.restoreInsertionPoint(insPt);
```
With my change to use the insertion point guard above
```suggestion
```

tblah reviewed (commented):
I have some minor suggestions on the code. Please wait for review from somebody with more familiarity with omp target things, and this is conditional on the design of the MLIR operation being approved.

TIFitis committed:
Address reviewer comments.

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2612,7 +2612,52 @@ static void
+  const auto &mapperName{std::get<std::optional<parser::Name>>(spec.t)};
+  const auto &varType{std::get<parser::TypeSpec>(spec.t)};
+  const auto &varName{std::get<parser::Name>(spec.t)};
+  assert(varType.declTypeSpec->category() ==
+             semantics::DeclTypeSpec::Category::TypeDerived &&
+         "Expected derived type");
```
Do you know if we have a semantics check for this, or if that should be added? Not a suggestion to do anything about that in this PR.

TIFitis replied:
I believe we have a semantics check in place for this.

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2612,7 +2612,52 @@ static void
+  // the DeclareMapperOp
+  mlir::OpBuilder::InsertionGuard guard(firOpBuilder);
+
+  firOpBuilder.setInsertionPointToStart(converter.getModuleOp().getBody());
```
Would adding each `declare mapper` at the top of the module cause issues in the case of "nested" declare mappers?

Let's say there's a derived type A for which a `declare mapper` exists and another derived type B holding an instance of A that also has a `declare mapper` which maps the instance of A using its `declare mapper`.

Following this approach would mean that the `omp.declare_mapper` for B, which is encountered later in the source code, would be placed on top and contain an `omp.map.info` with a reference to the `omp.declare_mapper` for A, which would be defined below.

This might not be an issue in MLIR (I'm not certain, but I believe module symbols are visible to the whole module), but if we end up translating `omp.declare_mapper` operations into functions, we may have to pay some attention into how we sort them later on.

TIFitis replied:
I have a test for this scenario in #124746.

skatrak replied:
Can you point me to the exact test you're referring to? I took a look at that PR and couldn't find any that checked this.

TIFitis replied:
`offload/test/offloading/fortran/target-custom-mapper.f90` uses a mapper within a mapper. Let me know if that's the scenario you were referring to here.

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2612,7 +2612,52 @@ static void
+  auto mlirType = converter.genType(varType.declTypeSpec->derivedTypeSpec());
+  auto declMapperOp = firOpBuilder.create<mlir::omp::DeclareMapperOp>(
+      loc, mapperNameStr, mlirType);
+  converter.getMLIRSymbolTable()->insert(declMapperOp);
```
Is this needed? Whenever we want to access the newly created operation by symbol, we can just query the MLIR module: `moduleOp.lookupSymbol<mlir::omp::DeclareMapperOp>(mangledName)`.

skatrak commented on flang/test/Lower/OpenMP/declare-mapper.f90:
```diff
@@ -0,0 +1,85 @@
+   end type
+   type(my_type2)        :: t
+   real                   :: x, y(nvals)
+   !CHECK-LABEL:omp.declare_mapper @_QQFdeclare_mapper_1my_type.default : !fir.type<_QFdeclare_mapper_1Tmy_type{num_vals:i32,values:!fir.box<!fir.heap<!fir.array<?xi32>>>}> {
```
Nit: I think it would make this test much simpler to understand and update in the future if we gave these long type strings and declare mapper symbols a name and reused them.
```suggestion
   !CHECK-LABEL:omp.declare_mapper @[[MY_TYPE_MAPPER:_QQFdeclare_mapper_1my_type\.default]] : [[MY_TYPE:!fir\.type<_QFdeclare_mapper_1Tmy_type{num_vals:i32,values:!fir\.box<!fir\.heap<!fir\.array<?xi32>>>}>]] {
```

skatrak reviewed (commented):
This generally looks good, I just have a few small comments.

Also, I see the corresponding code to match `omp.declare_mapper` operations with their uses in `map` clauses is missing at the moment, so this lowering only generates "orphan" operations. Is that coming as a separate patch?

TIFitis committed:
Addressed reviewer comments.

TIFitis committed:
Fix build issue.

skatrak commented on flang/test/Lower/OpenMP/declare-mapper.f90:
```diff
@@ -0,0 +1,85 @@
+   !CHECK:        %[[VAL_14:.*]] = omp.map.info var_ptr(%[[VAL_1]]#1 : !fir.ref<[[MY_TYPE]]>, [[MY_TYPE]]) map_clauses(tofrom) capture(ByRef) members(%[[VAL_9]], %[[VAL_12]] : [3], [1] : !fir.ref<!fir.array<250xf32>>, !fir.ref<!fir.type<_QFdeclare_mapper_2Tmy_type{num_vals:i32,values:!fir.box<!fir.heap<!fir.array<?xi32>>>}>>) -> !fir.ref<[[MY_TYPE]]> {name = "v", partial_map = true}
+   !CHECK:        omp.declare_mapper.info map_entries(%[[VAL_14]], %[[VAL_9]], %[[VAL_12]] : !fir.ref<[[MY_TYPE]]>, !fir.ref<!fir.array<250xf32>>, !fir.ref<!fir.type<_QFdeclare_mapper_2Tmy_type{num_vals:i32,values:!fir.box<!fir.heap<!fir.array<?xi32>>>}>>)
+   !CHECK:      }
+   !$omp declare mapper (my_mapper : my_type2 :: v) map (v%arr) map (alloc : v%temp)
```
It would probably be good to add another one of these tests where there's a declare mapper for `my_type` and a declare mapper for `my_type2` that uses the mapper defined for the first for one of its fields.

TIFitis replied:
I've introduced this test as part of #121001 as we would need map clause lowering support for mappers as well for this.

TIFitis committed:
Change isa to isa_and_present.

skatrak reviewed (approved):
LGTM, thank you.
